File: dotnet/src/Pactman.NonprofitCheckPlus/Models/Nonprofit.cs
using System.Collections.Generic;
using System.Text.Json;

namespace Pactman.NonprofitCheckPlus.Models
{
    /// <summary>
    /// A nonprofit record as returned by the US nonprofit check endpoints.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Source-specific findings are flat on this object, prefixed by source
    /// (<c>pub78_*</c>, <c>bmf_*</c>, <c>ofac_*</c>, and the revocation fields for the
    /// IRS Automatic Revocation of Exemption list). See <see cref="NonprofitCheckPlus.Sources"/>
    /// for grouped views.
    /// </para>
    /// <para>
    /// Every field is optional: the API omits fields it has no data for. Reading one
    /// it did not return yields <see langword="null"/>; ask <see cref="DataObject.Has(string)"/>
    /// to tell the two apart. Fields introduced by a newer API version than this SDK
    /// knows about are readable through <see cref="DataObject.Get(string)"/>.
    /// </para>
    /// <para>
    /// Declared here is what the production API returns. Some deployments serve
    /// additional source fields — the BMF address (<c>bmf_city</c>, <c>bmf_state</c>,
    /// <c>bmf_street_address</c>), <c>bmf_source_pf_filing_req_cd</c>,
    /// <c>bmf_deductability_text</c>, <c>pub78_source_org_type_1..3</c>,
    /// <c>ofac_list_published_date</c> and <c>aroe_list_published_date</c>. They are not
    /// declared because production does not return them; when it does, they stay readable
    /// through <see cref="DataObject.Get(string)"/>, and this package will declare them in
    /// a release of its own.
    /// </para>
    /// </remarks>
    public sealed class Nonprofit : DataObject
    {
        /// <summary>Initializes the record from a decoded JSON object.</summary>
        /// <param name="element">The organization as the API sent it.</param>
        public Nonprofit(JsonElement element)
            : base(element)
        {
        }

        /// <summary>Initializes the record from a set of wire fields.</summary>
        /// <param name="fields">The organization's fields, in document order.</param>
        public Nonprofit(IEnumerable<KeyValuePair<string, JsonElement>> fields)
            : base(fields)
        {
        }

        /// <summary>Public Pactman profile URL for the organization.</summary>
        public string? PactmanOrgUrl => GetString("pactman_org_url");

        /// <summary>When Pactman last modified this organization's record.</summary>
        public string? OrganizationInfoLastModified => GetString("organization_info_last_modified");

        /// <summary>The organization's EIN, as nine digits.</summary>
        public string? Ein => GetString("ein");

        /// <summary>The organization's legal name.</summary>
        public string? OrganizationName => GetString("organization_name");

        /// <summary>The organization's "also known as" name.</summary>
        public string? OrganizationNameAka => GetString("organization_name_aka");

        /// <summary>First line of the organization's address.</summary>
        public string? AddressLine1 => GetString("address_line1");

        /// <summary>Second line of the organization's address.</summary>
        public string? AddressLine2 => GetString("address_line2");

        /// <summary>The organization's city.</summary>
        public string? City => GetString("city");

        /// <summary>The organization's two-letter state code.</summary>
        public string? State => GetString("state");

        /// <summary>The organization's state, spelled out.</summary>
        public string? StateName => GetString("state_name");

        /// <summary>The organization's ZIP code.</summary>
        public string? Zip => GetString("zip");

        /// <summary>The IRS filing requirement code.</summary>
        public string? FilingReqCode => GetString("filing_req_code");

        /// <summary>The Publication 78 note for organizations the IRS treats as churches.</summary>
        public string? Pub78ChurchMessage => GetString("pub78_church_message");

        /// <summary>The organization's name as Publication 78 records it.</summary>
        public string? Pub78OrganizationName => GetString("pub78_organization_name");

        /// <summary>The EIN as Publication 78 records it.</summary>
        public string? Pub78Ein => GetString("pub78_ein");

        /// <summary>Whether the organization appears in Publication 78.</summary>
        public bool? Pub78Verified => GetBoolean("pub78_verified");

        /// <summary>The city as Publication 78 records it.</summary>
        public string? Pub78City => GetString("pub78_city");

        /// <summary>The state as Publication 78 records it.</summary>
        public string? Pub78State => GetString("pub78_state");

        /// <summary>The Publication 78 deductibility indicator.</summary>
        public string? Pub78Indicator => GetString("pub78_indicator");

        /// <summary>The date of the most recent Publication 78 extract behind this record.</summary>
        public string? MostRecentPub78 => GetString("most_recent_pub78");

        /// <summary>The Business Master File note for organizations the IRS treats as churches.</summary>
        public string? BmfChurchMessage => GetString("bmf_church_message");

        /// <summary>The organization's name as the Business Master File records it.</summary>
        public string? BmfOrganizationName => GetString("bmf_organization_name");

        /// <summary>The EIN as the Business Master File records it.</summary>
        public string? BmfEin => GetString("bmf_ein");

        /// <summary>Whether the organization appears in the Business Master File.</summary>
        public bool? BmfStatus => GetBoolean("bmf_status");

        /// <summary>The IRC subsection under which the organization is exempt.</summary>
        public string? BmfSubsection => GetString("bmf_subsection");

        /// <summary>The date of the most recent Business Master File extract behind this record.</summary>
        public string? MostRecentBmf => GetString("most_recent_bmf");

        /// <summary>The subsection, spelled out.</summary>
        public string? SubsectionDescription => GetString("subsection_description");

        /// <summary>The IRS foundation code.</summary>
        public string? FoundationCode => GetString("foundation_code");

        /// <summary>The foundation code, spelled out.</summary>
        public string? FoundationCodeDescription => GetString("foundation_code_description");

        /// <summary>The IRS foundation type code.</summary>
        public string? FoundationTypeCode => GetString("foundation_type_code");

        /// <summary>The foundation type, spelled out.</summary>
        public string? FoundationTypeDescription => GetString("foundation_type_description");

        /// <summary>The organization's 509(a) status.</summary>
        public string? Foundation509aStatus => GetString("foundation_509a_status");

        /// <summary>The month of the IRS ruling.</summary>
        public string? RulingMonth => GetString("ruling_month");

        /// <summary>The year of the IRS ruling.</summary>
        public string? RulingYear => GetString("ruling_year");

        /// <summary>The group exemption number, when the organization holds one.</summary>
        public string? GroupExemption => GetString("group_exemption");

        /// <summary>The IRS exempt status code.</summary>
        public string? ExemptStatusCode => GetString("exempt_status_code");

        /// <summary>
        /// The OFAC Specially Designated Nationals finding.
        /// </summary>
        /// <remarks>
        /// This is prose, not a flag. The API does not currently return a boolean match
        /// indicator, and this SDK does not invent one by matching on the wording.
        /// </remarks>
        public string? OfacStatus => GetString("ofac_status");

        /// <summary>The IRS automatic revocation code, when the exemption was revoked.</summary>
        public string? RevocationCode => GetString("revocation_code");

        /// <summary>The date the exemption was automatically revoked.</summary>
        public string? RevocationDate => GetString("revocation_date");

        /// <summary>The date the exemption was reinstated, when it was.</summary>
        public string? ReinstatementDate => GetString("reinstatement_date");

        /// <summary>True when the IRS Business Master File and Publication 78 records disagree.</summary>
        public bool? IrsBmfPub78Conflict => GetBoolean("irs_bmf_pub78_conflict");

        /// <summary>The date Pactman compiled this report.</summary>
        public string? ReportDate => GetString("report_date");

        /// <summary>
        /// The Publication 78 deductibility entries, as the API sent them.
        /// </summary>
        /// <remarks>
        /// <para>
        /// Returns an empty list when the field was absent or null, so it is always safe
        /// to enumerate; ask <c>Has("organization_types")</c> when the difference matters.
        /// </para>
        /// <para>
        /// The API can also send a null in the list, for a Publication 78 row it cannot
        /// resolve. Those are dropped rather than handed on, so every entry this returns
        /// is a real object and the positions are renumbered from zero — read
        /// <c>Get("organization_types")</c> when the gaps themselves matter.
        /// </para>
        /// </remarks>
        public IReadOnlyList<OrganizationType> OrganizationTypes =>
            OrganizationType.ListFrom(GetElement("organization_types"));
    }
}

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/models/Nonprofit.java
package org.pactman.nonprofitcheckplus.models;

import java.util.List;
import java.util.Map;

/**
 * A nonprofit record as returned by the US nonprofit check endpoints.
 *
 * <p>Source-specific findings are flat on this object, prefixed by source
 * ({@code pub78_*}, {@code bmf_*}, {@code ofac_*}, and the revocation fields for
 * the IRS Automatic Revocation of Exemption list). See
 * {@link org.pactman.nonprofitcheckplus.Sources} for grouped views.
 *
 * <p>Every field is optional: the API omits fields it has no data for. Reading
 * one it did not return yields {@code null}; ask {@link DataObject#has(String)}
 * to tell the two apart. Fields introduced by a newer API version than this SDK
 * knows about are readable through {@link DataObject#get(String)}.
 *
 * <p>Declared here is what the production API returns. Some deployments serve
 * additional source fields — the BMF address ({@code bmf_city},
 * {@code bmf_state}, {@code bmf_street_address}),
 * {@code bmf_source_pf_filing_req_cd}, {@code bmf_deductability_text},
 * {@code pub78_source_org_type_1..3}, {@code ofac_list_published_date} and
 * {@code aroe_list_published_date}. They are not declared because production
 * does not return them; when it does, they stay readable through
 * {@link DataObject#get(String)}, and this package will declare them in a
 * release of its own.
 */
public final class Nonprofit extends DataObject {

    /**
     * Initializes the record from a decoded JSON object.
     *
     * @param fields the organization as the API sent it.
     */
    public Nonprofit(Map<String, Object> fields) {
        super(fields);
    }

    /**
     * Public Pactman profile URL for the organization.
     *
     * @return the URL, or {@code null} when the API returned none.
     */
    public String getPactmanOrgUrl() {
        return getString("pactman_org_url");
    }

    /**
     * When Pactman last modified this organization's record.
     *
     * @return the timestamp as the API formats it, or {@code null}.
     */
    public String getOrganizationInfoLastModified() {
        return getString("organization_info_last_modified");
    }

    /**
     * The organization's EIN, as nine digits.
     *
     * @return the EIN, or {@code null} when the API returned none.
     */
    public String getEin() {
        return getString("ein");
    }

    /**
     * The organization's legal name.
     *
     * @return the name, or {@code null} when the API returned none.
     */
    public String getOrganizationName() {
        return getString("organization_name");
    }

    /**
     * The organization's "also known as" name.
     *
     * @return the alternate name, or {@code null} when the API returned none.
     */
    public String getOrganizationNameAka() {
        return getString("organization_name_aka");
    }

    /**
     * First line of the organization's address.
     *
     * @return the line, or {@code null} when the API returned none.
     */
    public String getAddressLine1() {
        return getString("address_line1");
    }

    /**
     * Second line of the organization's address.
     *
     * @return the line, or {@code null} when the API returned none.
     */
    public String getAddressLine2() {
        return getString("address_line2");
    }

    /**
     * The organization's city.
     *
     * @return the city, or {@code null} when the API returned none.
     */
    public String getCity() {
        return getString("city");
    }

    /**
     * The organization's two-letter state code.
     *
     * @return the state code, or {@code null} when the API returned none.
     */
    public String getState() {
        return getString("state");
    }

    /**
     * The organization's state, spelled out.
     *
     * @return the state name, or {@code null} when the API returned none.
     */
    public String getStateName() {
        return getString("state_name");
    }

    /**
     * The organization's ZIP code.
     *
     * @return the ZIP, or {@code null} when the API returned none.
     */
    public String getZip() {
        return getString("zip");
    }

    /**
     * The IRS filing requirement code.
     *
     * @return the code, or {@code null} when the API returned none.
     */
    public String getFilingReqCode() {
        return getString("filing_req_code");
    }

    /**
     * The Publication 78 note for organizations the IRS treats as churches.
     *
     * @return the note, or {@code null} when the API returned none.
     */
    public String getPub78ChurchMessage() {
        return getString("pub78_church_message");
    }

    /**
     * The organization's name as Publication 78 records it.
     *
     * @return the name, or {@code null} when the API returned none.
     */
    public String getPub78OrganizationName() {
        return getString("pub78_organization_name");
    }

    /**
     * The EIN as Publication 78 records it.
     *
     * @return the EIN, or {@code null} when the API returned none.
     */
    public String getPub78Ein() {
        return getString("pub78_ein");
    }

    /**
     * Whether the organization appears in Publication 78.
     *
     * @return the flag, or {@code null} when the API returned none.
     */
    public Boolean getPub78Verified() {
        return getBoolean("pub78_verified");
    }

    /**
     * The city as Publication 78 records it.
     *
     * @return the city, or {@code null} when the API returned none.
     */
    public String getPub78City() {
        return getString("pub78_city");
    }

    /**
     * The state as Publication 78 records it.
     *
     * @return the state, or {@code null} when the API returned none.
     */
    public String getPub78State() {
        return getString("pub78_state");
    }

    /**
     * The Publication 78 deductibility indicator.
     *
     * @return the indicator, or {@code null} when the API returned none.
     */
    public String getPub78Indicator() {
        return getString("pub78_indicator");
    }

    /**
     * The Publication 78 deductibility entries, as the API sent them.
     *
     * <p>Empty when the field was absent or null, so it is always safe to
     * iterate; ask {@code has("organization_types")} when the difference matters.
     *
     * @return the entries.
     */
    public List<OrganizationType> getOrganizationTypes() {
        return OrganizationType.listFrom(get("organization_types"));
    }

    /**
     * The date of the most recent Publication 78 extract behind this record.
     *
     * @return the date as the API formats it, or {@code null}.
     */
    public String getMostRecentPub78() {
        return getString("most_recent_pub78");
    }

    /**
     * The Business Master File note for organizations the IRS treats as churches.
     *
     * @return the note, or {@code null} when the API returned none.
     */
    public String getBmfChurchMessage() {
        return getString("bmf_church_message");
    }

    /**
     * The organization's name as the Business Master File records it.
     *
     * @return the name, or {@code null} when the API returned none.
     */
    public String getBmfOrganizationName() {
        return getString("bmf_organization_name");
    }

    /**
     * The EIN as the Business Master File records it.
     *
     * @return the EIN, or {@code null} when the API returned none.
     */
    public String getBmfEin() {
        return getString("bmf_ein");
    }

    /**
     * Whether the organization appears in the Business Master File.
     *
     * @return the flag, or {@code null} when the API returned none.
     */
    public Boolean getBmfStatus() {
        return getBoolean("bmf_status");
    }

    /**
     * The IRC subsection under which the organization is exempt.
     *
     * @return the subsection, or {@code null} when the API returned none.
     */
    public String getBmfSubsection() {
        return getString("bmf_subsection");
    }

    /**
     * The date of the most recent Business Master File extract behind this record.
     *
     * @return the date as the API formats it, or {@code null}.
     */
    public String getMostRecentBmf() {
        return getString("most_recent_bmf");
    }

    /**
     * The subsection, spelled out.
     *
     * @return the description, or {@code null} when the API returned none.
     */
    public String getSubsectionDescription() {
        return getString("subsection_description");
    }

    /**
     * The IRS foundation code.
     *
     * @return the code, or {@code null} when the API returned none.
     */
    public String getFoundationCode() {
        return getString("foundation_code");
    }

    /**
     * The foundation code, spelled out.
     *
     * @return the description, or {@code null} when the API returned none.
     */
    public String getFoundationCodeDescription() {
        return getString("foundation_code_description");
    }

    /**
     * The IRS foundation type code.
     *
     * @return the code, or {@code null} when the API returned none.
     */
    public String getFoundationTypeCode() {
        return getString("foundation_type_code");
    }

    /**
     * The foundation type, spelled out.
     *
     * @return the description, or {@code null} when the API returned none.
     */
    public String getFoundationTypeDescription() {
        return getString("foundation_type_description");
    }

    /**
     * The organization's 509(a) status.
     *
     * @return the status, or {@code null} when the API returned none.
     */
    public String getFoundation509aStatus() {
        return getString("foundation_509a_status");
    }

    /**
     * The month of the IRS ruling.
     *
     * @return the month, or {@code null} when the API returned none.
     */
    public String getRulingMonth() {
        return getString("ruling_month");
    }

    /**
     * The year of the IRS ruling.
     *
     * @return the year, or {@code null} when the API returned none.
     */
    public String getRulingYear() {
        return getString("ruling_year");
    }

    /**
     * The group exemption number, when the organization holds one.
     *
     * @return the number, or {@code null} when the API returned none.
     */
    public String getGroupExemption() {
        return getString("group_exemption");
    }

    /**
     * The IRS exempt status code.
     *
     * @return the code, or {@code null} when the API returned none.
     */
    public String getExemptStatusCode() {
        return getString("exempt_status_code");
    }

    /**
     * The OFAC Specially Designated Nationals finding.
     *
     * <p>This is prose, not a flag. The API does not currently return a boolean
     * match indicator, and this SDK does not invent one by matching on the
     * wording. Read it, or present it to a reviewer.
     *
     * @return the finding, or {@code null} when the API returned none.
     */
    public String getOfacStatus() {
        return getString("ofac_status");
    }

    /**
     * The IRS automatic revocation code, when the exemption was revoked.
     *
     * @return the code, or {@code null} when the API returned none.
     */
    public String getRevocationCode() {
        return getString("revocation_code");
    }

    /**
     * The date the exemption was automatically revoked.
     *
     * @return the date as the API formats it, or {@code null}.
     */
    public String getRevocationDate() {
        return getString("revocation_date");
    }

    /**
     * The date the exemption was reinstated, when it was.
     *
     * @return the date as the API formats it, or {@code null}.
     */
    public String getReinstatementDate() {
        return getString("reinstatement_date");
    }

    /**
     * True when the IRS Business Master File and Publication 78 records disagree.
     *
     * @return the flag, or {@code null} when the API returned none.
     */
    public Boolean getIrsBmfPub78Conflict() {
        return getBoolean("irs_bmf_pub78_conflict");
    }

    /**
     * The date Pactman compiled this report.
     *
     * @return the date as the API formats it, or {@code null}.
     */
    public String getReportDate() {
        return getString("report_date");
    }
}
